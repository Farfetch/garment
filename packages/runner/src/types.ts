export type File = TextFile | JsonFile | SourceFile | BinaryFile;

export interface JsonFile extends BaseFile {
  type: 'json';
  data: object;
}

export interface SourceFile extends BaseFile {
  type: 'source';
  data: string;
  map?: any;
}

export interface TextFile extends BaseFile {
  type: 'text';
  data: string;
}

export interface BinaryFile extends BaseFile {
  type: 'binary';
  data: Buffer;
}

export interface BaseFile {
  path: string;
  absolutePath?: string;
  baseDir?: string;
  skipWrite?: boolean;
}
