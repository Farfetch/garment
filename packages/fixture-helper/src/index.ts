export * from './executeRunner';
export * from './initFixtureHelper';
export * from './resolver';
import normalizePath = require('normalize-path');
import isValidPath = require('is-valid-path');

const containsValidPath = (value: string): boolean =>
  value.split(/\s/).some((s) => isValidPath(s));

export const normalizeContainedPath = (value: string): string => {
  return value
    .split(/(\s)/)
    .map((s) => {
      if (isValidPath(s)) {
        return normalizePath(s);
      } else {
        return s;
      }
    })
    .join('');
};

export function replaceTestPath<T>(
  obj: T,
  testPath: string,
  replaceWith = '/test_path'
): T {
  const normalizedTestPathRegEx = new RegExp(normalizePath(testPath), 'g');
  return JSON.parse(
    JSON.stringify(obj, (_, value) =>
      typeof value === 'string' && containsValidPath(value)
        ? normalizeContainedPath(value).replace(
            normalizedTestPathRegEx,
            replaceWith
          )
        : value
    )
  );
}
