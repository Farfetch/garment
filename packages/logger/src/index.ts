import chalk from 'chalk';
import stripAnsi from 'strip-ansi';

const levelsMap = {
  silly: 0,
  debug: 1,
  info: 2,
  success: 3,
  warn: 4,
  error: 5,
  silent: 6,
};

export type Level = keyof typeof levelsMap;

const logTypeToTextMap: { [key in Level]?: string } = {
  silly: chalk.keyword('purple')('silly'),
  debug: chalk.keyword('cyan')('debug'),
  info: chalk.keyword('lightblue')('info'),
  warn: chalk.keyword('yellow')('warning'),
  error: chalk.keyword('darkred')('error'),
  success: chalk.keyword('lime')('success'),
};

export function shouldPrint(level: Level, currentLevel: Level) {
  return levelsMap[level] >= levelsMap[currentLevel];
}

export const levels = Object.keys(levelsMap);

export class Logger {
  public static print(
    options: {
      prefix?: string;
      scope?: string;
      level?: Level;
      scopeColor?: string;
      dontLog?: boolean;
    } = {},
    ...args: any[]
  ) {
    const meta: string[] = [];
    const { prefix, scope, level, scopeColor = 'magenta', dontLog } = options;
    if (prefix) {
      meta.push(chalk.grey(prefix));
    }
    if (scope) {
      meta.push(chalk.keyword(scopeColor)(scope));
    }
    if (level) {
      meta.push(logTypeToTextMap[level] || level);
    }
    const joinedMeta = meta.join(' ');

    const indentedArgs = args
      .map((arg) => (Object(arg) !== arg ? arg : JSON.stringify(arg)))
      .join(' ')
      .split(/\n/)
      .join('\n' + ' '.repeat(stripAnsi(joinedMeta).length + 1));

    const consoleLogArgs: string[] = [joinedMeta, indentedArgs];
    if (!dontLog) {
      console.log(...consoleLogArgs);
    }
    return consoleLogArgs;
  }

  interceptors: ((entry: { level: Level; args: any[] }) => any)[] = [];

  constructor(
    public readonly prefix: string | undefined,
    public readonly name: string,
    public level: Level = 'info',
    public readonly color?: string
  ) {}

  public silly(...args: any[]) {
    this.display('silly', ...args);
  }

  public debug(...args: any[]) {
    this.display('debug', ...args);
  }

  public info(...args: any[]) {
    this.display('info', ...args);
  }

  public warn(...args: any[]) {
    this.display('warn', ...args);
  }

  public error(...args: any[]) {
    this.display('error', chalk.keyword('red')(...args));
  }

  public success(...args: any[]) {
    this.display('success', ...args);
  }

  public log(...args: any[]) {
    Logger.print({}, ...args);
  }

  public createChild(scope: string, color?: string) {
    return new Logger(this.prefix, scope, this.level, color);
  }

  public setLevel(level: Level) {
    this.level = level;
  }

  private display(level: Level, ...args: any[]) {
    if (
      !this.interceptors.every((fn) => {
        const result = fn({ level, args });
        return typeof result === 'boolean' ? result : true;
      })
    ) {
      return;
    }
    if (shouldPrint(level, this.level)) {
      Logger.print(
        {
          prefix: this.prefix,
          scope: this.name,
          level,
          scopeColor: this.color,
        },
        ...args
      );
    }
  }
}

export default new Logger('', '', 'info');
