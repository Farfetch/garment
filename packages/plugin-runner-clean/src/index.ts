import { defineRunner } from '@garment/runner';
import del from 'del';

/**
 * Deletes files and folders using glob patterns.
 */
export interface CleanRunnerOptions {}

export default defineRunner(async function (ctx) {
  ctx.logger.info(`Removing files`);

  ctx.input.forEach(async ({ path, baseDir }) => {
    if (baseDir) {
      await del(path, {
        cwd: baseDir,
      });
    } else {
      throw new Error(`Unable to delete "${path}". baseDir is empty`);
    }
  });

  ctx.logger.success(`Files removed`);
});
