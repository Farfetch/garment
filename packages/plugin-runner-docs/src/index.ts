import { defineRunner, defineOptions } from '@garment/runner';
import * as Path from 'path';
import { Project, SourceFile } from 'ts-morph';

export default defineRunner(
  defineOptions({
    template: ['path', 'Path to the template'],
    filename: ['string?', 'Name of the output file', 'README.md']
  }),
  async ctx => {
    const { template, filename } = ctx.options;
    const packageName = ctx.project.nodePackage?.name;

    const templateContent = ctx.renderTemplate(
      ctx.fs.readFileSync(template, 'utf8'),
      { packageName }
    );

    const templateLines = templateContent.split('\n');
    const updatedTemplateLines = [];

    const project = new Project({});
    const sourceFiles: Record<string, SourceFile> = {};

    let ignoreLine = false;

    for (const line of templateLines) {
      const match = /<!--\s*(.*)\s*-->/.exec(line);
      if (match) {
        const [op, ...args] = match[1].split(/\s+/);

        if (op === 'api') {
          const [filename, declarationName] = args;

          const sourceFile = getSourceByFilename(filename);

          const declaration =
            declarationName === 'firstInterface'
              ? sourceFile.getInterfaces()[0]
              : sourceFile.getInterfaceOrThrow(declarationName);

          const allResults = [];

          for (const member of declaration.getMembers()) {
            let result = [];
            const name = (member as any).getName();
            let type = member.getType().getText();
            let defaultValue;
            let comment;

            const [jsdoc] = member.getJsDocs();
            if (jsdoc) {
              comment = jsdoc.getDescription().trim();
              const tags: Record<string, string> = {};
              for (const tag of jsdoc.getTags()) {
                tags[tag.getTagName()] = tag.getComment()?.trim() ?? '';
              }

              if (tags.format === 'path') {
                type = 'path';
              }

              defaultValue = tags.defaultDescription || tags.default;
            }

            result.push(`\`${name}: ${type}\``);
            if (defaultValue) {
              result.push(`**Default:** \`${defaultValue}\``);
            }
            if (comment) {
              result.push(comment);
            }

            allResults.push(result.join('\n\n'));
          }

          ignoreLine = true;

          updatedTemplateLines.push(line);

          allResults.forEach((result, index) => {
            updatedTemplateLines.push(result);
            if (index === allResults.length - 1) {
              updatedTemplateLines.push('\n');
            } else {
              updatedTemplateLines.push('\n---\n');
            }
          });
        } else if (op === 'example' || op === 'description') {
          const [filename, declarationName] = args;

          const sourceFile = getSourceByFilename(filename);

          const declaration =
            declarationName === 'firstInterface'
              ? sourceFile.getInterfaces()[0]
              : sourceFile.getInterfaceOrThrow(declarationName);

          const allResults = [];

          const [jsdoc] = declaration.getJsDocs();
          if (jsdoc) {
            if (op === 'example') {
              const tags: Record<string, string> = {};
              for (const tag of jsdoc.getTags()) {
                tags[tag.getTagName()] = tag.getComment()?.trim() ?? '';
              }

              if (tags.example) {
                let parsed = tags.example;
                try {
                  parsed = JSON.stringify(JSON.parse(parsed), null, 4);
                } catch (error) {}

                allResults.push('```json');
                allResults.push(parsed);
                allResults.push('```\n');
              }
            } else if (op === 'description') {
              const description = jsdoc.getDescription().trim();
              allResults.push(description);
              allResults.push('\n');
            }
          }

          ignoreLine = true;

          updatedTemplateLines.push(line, ...allResults);
        } else if (op === 'installation') {
          updatedTemplateLines.push(line);
          if (packageName) {
            updatedTemplateLines.push(`\`npm i ${packageName}\`\n`);
            updatedTemplateLines.push(`or\n`);
            updatedTemplateLines.push(`\`yarn add ${packageName}\`\n`);
          }

          ignoreLine = true;
        } else if (op === 'break') {
          ignoreLine = false;
        }
      } else if (line[0] === '#') {
        ignoreLine = false;
      }

      if (!ignoreLine) {
        updatedTemplateLines.push(line);
      }
    }

    return ctx.file.text(filename, updatedTemplateLines.join('\n'));

    function getSourceByFilename(filename: string) {
      const resolvedFilename = Path.resolve(ctx.project.fullPath, filename);
      if (!sourceFiles[resolvedFilename]) {
        sourceFiles[resolvedFilename] = project.addSourceFileAtPath(
          resolvedFilename
        );
      }
      return sourceFiles[resolvedFilename];
    }
  }
);
