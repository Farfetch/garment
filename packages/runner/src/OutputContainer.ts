import { Level, Logger } from '@garment/logger';
import { File } from './types';
import { createHash } from 'crypto';
import * as Path from 'path';

export interface CacheProvider<T = any> {
  has(id: string): Promise<boolean> | boolean;
  get(id: string): Promise<T> | T;
  set(id: string, content: T): Promise<void> | void;
}
export interface LogEntry {
  level: Level;
  content: any[];
  scope: string;
}

export class OutputContainer {
  static isOutputContainer(obj: any): obj is OutputContainer {
    return (<OutputContainer>obj)?.isOutputContainer ?? false;
  }

  private readonly isOutputContainer = true;

  logger: Logger;
  logs: LogEntry[] = [];
  dependencies: string[];
  files: File[] = [];
  hash: string;
  target: string;

  targetBaseDir: string;

  constructor(
    public readonly cacheProvider: CacheProvider,
    target: string | File,
    public readonly cacheKeys: string[],
    dependencies: (string | File)[] = []
  ) {
    const hash = createHash('md5');
    for (const cacheKey of this.cacheKeys) {
      hash.update(JSON.stringify(cacheKey));
    }
    this.hash = hash.digest('hex');

    if (typeof target === 'string') {
      this.target = target;
      this.targetBaseDir = Path.dirname(target);
    } else {
      this.target = target.absolutePath ?? target.path;
      this.targetBaseDir = target.baseDir ?? Path.dirname(this.target);
    }

    this.dependencies = dependencies.map((dep) =>
      typeof dep === 'string' ? dep : dep.absolutePath ?? dep.path
    );

    this.logger = new Logger('', '', 'silly');
    this.logger.interceptors.push(({ level, args }) => {
      this.logs.push({
        level,
        content: args,
        scope: '    ',
      });
      return false; // creating logger instance which doesn't output to the console
    });
  }

  add(file: File) {
    this.files.push(file);
  }

  get isNotCached() {
    return !this.cacheProvider.has(this.hash);
  }

  async syncWithCache() {
    if (await this.cacheProvider.has(this.hash)) {
      const cacheEntry = await this.cacheProvider.get(this.hash);
      if (cacheEntry) {
        this.files = cacheEntry.output;
        this.logs = cacheEntry.logs;
      }
    } else {
      await this.cacheProvider.set(this.hash, {
        logs: this.logs,
        output: this.files,
      });
    }
  }

  toJSON() {
    const { cacheKeys, dependencies, logs, files } = this;
    return {
      cacheKeys,
      dependencies,
      logs,
      files,
    };
  }

  hasErrorsOrWarnings() {
    return this.logs.some(
      (entry) => entry.level === 'error' || entry.level === 'warn'
    );
  }
}
