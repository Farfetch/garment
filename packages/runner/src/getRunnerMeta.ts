import { OutputContainer } from './OutputContainer';
import { validateSchema } from '@garment/schema-validator';
import * as Path from 'path';
import { File } from './types';
import { Context } from './getRunnerContext';

export interface RunnerConfig {
  handler: string;
  description: string;
  watcher?: string;
  batch?: boolean;
  cacheKeys?: string;
}

export interface RunnersConfig {
  runners: {
    [name: string]: RunnerConfig;
  };
}

export function getRunnerMeta(runnerName: string, cwd: string) {
  if (Path.isAbsolute(runnerName) || runnerName[0] === '.') {
    return {
      name: runnerName,
      version: 'internal',
      handlerPath: Path.resolve(cwd, runnerName),
      description: 'no description',
      batch: false,
    };
  }

  const [packageName, scriptName = 'default'] = runnerName.split(':');

  const namesToTry = [`@garment/plugin-runner-${packageName}`, packageName];

  for (const resolvedPackageName of namesToTry) {
    try {
      // First check if package has runners.json file. In this case use it to get runners information
      const packageJsonPath = require.resolve(
        `${resolvedPackageName}/package.json`
      );

      const runnersPackageJson = require(packageJsonPath) as {
        main: string;
        version: string;
      };

      const packagePath = Path.dirname(packageJsonPath);
      const resolveFromPackage = (path: string) =>
        Path.resolve(packagePath, path);

      try {
        const runnersJson: RunnersConfig = require(resolveFromPackage(
          'runners.json'
        ));

        const error = validateSchema(
          require('../schemas/runners.schema.json'),
          runnersJson,
          runnerName + 'Schema'
        );
        if (error) {
          throw error;
        }

        const runnerConfig = runnersJson.runners[scriptName];
        if (!runnerConfig) {
          throw new Error(
            `Runner "${scriptName}" is not found in "${resolvedPackageName}"`
          );
        }

        return {
          name: runnerName,
          version: runnersPackageJson.version,
          handlerPath: resolveFromPackage(runnerConfig.handler),
          description: runnerConfig.description,
          watcherPath: runnerConfig.watcher
            ? resolveFromPackage(runnerConfig.watcher)
            : undefined,
          batch: runnerConfig.batch || false,
          cacheKeysPath: runnerConfig.cacheKeys
            ? resolveFromPackage(runnerConfig.cacheKeys)
            : undefined,
        };
      } catch (e) {
        if (e.code !== 'MODULE_NOT_FOUND') {
          throw e;
        }
      }

      // Then check if package resolves to the main file
      const runnerHandlerPath = runnersPackageJson.main;

      return {
        name: resolvedPackageName,
        queryName: packageName,
        version: runnersPackageJson.version,
        handlerPath: resolveFromPackage(runnerHandlerPath),
        description: 'no description',
        batch: false,
      };
    } catch (e) {
      if (e.code !== 'MODULE_NOT_FOUND') {
        throw e;
      }
    }
  }

  throw new Error(
    `Runner "${packageName}" not found. Tried: ${namesToTry.join(', ')}`
  );
}

export type RunnerMeta = NonNullable<ReturnType<typeof getRunnerMeta>>;

export interface RunnerHandler<O> {
  (context: Context<O>):
    | File
    | OutputContainer
    | File[]
    | OutputContainer[]
    | Promise<File>
    | Promise<File[]>
    | Promise<OutputContainer>
    | Promise<OutputContainer[]>
    | Promise<void>;
}

export interface RunnerCacheKeys<O> {
  (context: Context<O>): Promise<string[]> | string[];
}

export function getHandler(runnerMeta: RunnerMeta) {
  return requireQuery(runnerMeta.handlerPath).handler as RunnerHandler<any>;
}

export function getWatcher(runnerMeta: RunnerMeta) {
  if (runnerMeta.watcherPath) {
    return requireQuery(runnerMeta.watcherPath).handler as RunnerHandler<any>;
  } else {
    throw new Error(`Runner "${runnerMeta.name}" doesn't support watch mode`);
  }
}

export function getCacheKeys({ cacheKeysPath }: RunnerMeta) {
  return cacheKeysPath
    ? (requireQuery(cacheKeysPath).cacheKeysFn as RunnerCacheKeys<any>)
    : (arg: any) => [];
}

export function getSchema(runnerMeta: RunnerMeta) {
  return requireQuery(runnerMeta.handlerPath)?.options?.schema ?? {};
}

export function validateOptions(runnerMeta: RunnerMeta, options: any) {
  const { outputDir, ...restOptions } = filterOptions(options);
  const error = validateSchema(
    getSchema(runnerMeta),
    restOptions,
    runnerMeta.name + 'Options'
  );
  if (error) {
    throw error;
  }
}

export function requireQuery<T = any>(query: string): T {
  const [path, exportName] = query.split('#');
  const exports = require(path);
  return exportName ? exports[exportName] : exports.default || exports;
}

function filterOptions<T extends { [key in keyof T]: any }>(obj: T) {
  const result: { [key: string]: any } = {};
  for (const [key, value] of Object.entries(obj)) {
    if (value) {
      result[key] = value;
    }
  }
  return result as T;
}
