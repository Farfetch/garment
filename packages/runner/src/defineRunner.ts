import { RunnerHandler, RunnerCacheKeys } from './getRunnerMeta';

type TypeMap = {
  'path?'?: string;
  path: string;
  'number?'?: number;
  number: number;
  'string?'?: string;
  string: string;
  'boolean?'?: boolean;
  boolean: boolean;
  'path[]?'?: string[];
  'path[]': string[];
  'number[]?'?: number[];
  'number[]': number[];
  'string[]?'?: string[];
  'string[]': string[];
  'boolean[]?'?: boolean[];
  'boolean[]': boolean[];
};

type TypeMapWithDefault = {
  'path?': string;
  'number?': number;
  'string?': string;
  'boolean?': boolean;
  'path[]?': string[];
  'number[]?': number[];
  'string[]?': string[];
  'boolean[]?': boolean[];
};

type TypeKeys = keyof TypeMap | keyof TypeMapWithDefault;

type OptionsArg = {
  [key: string]:
    | TypeKeys
    | [TypeKeys]
    | [TypeKeys, string]
    | ['path?', string, TypeMap['path']]
    | ['number?', string, TypeMap['number']]
    | ['string?', string, TypeMap['string']]
    | ['boolean?', string, TypeMap['boolean']];
  // | OptionsArg;
};

export type RunnerOptions<O extends OptionsArg> = {
  [key in keyof O]: O[key][0] extends TypeKeys
    ? O[key] extends [any, string, any]
      ? TypeMapWithDefault[O[key][0]]
      : TypeMap[O[key][0]]
    : O[key] extends TypeKeys
    ? TypeMap[O[key]]
    : O[key] extends OptionsArg
    ? RunnerOptions<O[key]>
    : never;
};

export type OptionsSchema<T> = {
  schema: any;
  type: T;
};

export function defineOptions<T extends OptionsArg>(
  opts: T
): OptionsSchema<RunnerOptions<T>> {
  const schema = {
    properties: {} as {
      [prop: string]: {
        type: string;
        description?: string;
        format?: string;
        default?: any;
        items?: any;
      };
    },
    required: [] as string[],
    additionalProperties: false
  };
  Object.entries(opts).forEach(([key, value]) => {
    const [typeWithModifier, description, defaultValue] = Array.isArray(value)
      ? value
      : [value];
    const [, type, isArray, isOptional] =
      typeWithModifier.match(/([a-z]*)(\[\])?(\?)?$/) || [];
    schema.properties[key] = isArray
      ? {
          type: 'array',
          items: {
            type,
            description,
            default: defaultValue,
            ...(type === 'path' && { type: 'string', format: 'path' })
          }
        }
      : {
          type,
          description,
          default: defaultValue,
          ...(type === 'path' && { type: 'string', format: 'path' })
        };
    if (!isOptional) {
      schema.required.push(key);
    }
  });
  return {
    schema,
    type: {} as RunnerOptions<T>
  };
}

export function defineOptionsFromJSONSchema<T>(schema: any): OptionsSchema<T> {
  return {
    schema,
    type: {} as T
  };
}

export function defineRunner(
  handler: RunnerHandler<Record<string, any>>
): { options: OptionsSchema<Record<string, any>>; handler: typeof handler };
export function defineRunner<O>(
  options: OptionsSchema<O>,
  handler: RunnerHandler<O>
): { options: typeof options; handler: typeof handler };
export function defineRunner<O>(
  paramOne: OptionsSchema<O> | RunnerHandler<O>,
  handler?: RunnerHandler<O>
) {
  return typeof paramOne === 'function'
    ? { options: {}, handler: paramOne }
    : { options: paramOne, handler };
}

export function defineCacheKeys<O>(
  options: OptionsSchema<O> | null,
  cacheKeysFn: RunnerCacheKeys<O>
) {
  return { options, cacheKeysFn };
}

export function defineWatch<O>(
  options: OptionsSchema<O> | {},
  handler: RunnerHandler<O>
) {
  return { options, handler };
}
