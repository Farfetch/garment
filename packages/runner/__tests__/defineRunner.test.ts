import { defineOptions } from '../src/defineRunner';

describe('defineRunner', () => {
  test('returns correct schema', () => {
    expect(
      defineOptions({ foo: ['string'], bar: ['number?'], configFile: ['path'] })
        .schema
    ).toMatchSnapshot();
  });
});
