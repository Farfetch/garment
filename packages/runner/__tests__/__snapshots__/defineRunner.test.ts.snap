// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineRunner returns correct schema 1`] = `
Object {
  "additionalProperties": false,
  "properties": Object {
    "bar": Object {
      "default": undefined,
      "description": undefined,
      "type": "number",
    },
    "configFile": Object {
      "default": undefined,
      "description": undefined,
      "format": "path",
      "type": "string",
    },
    "foo": Object {
      "default": undefined,
      "description": undefined,
      "type": "string",
    },
  },
  "required": Array [
    "foo",
    "configFile",
  ],
}
`;
